alvr/sockets/src/stream_socket/tcp.rs

```diff
@@ -0,0 +1,229 @@
+
+    if let Some(ip) = server_ip
+        && server_address.ip() != ip
+    {
+        con_bail!(
+            "Connected to wrong client: Expected: {ip}, Found {}",
+            server_address.ip()
+        );
+    }
+
+    socket.set_read_timeout(Some(timeout)).to_con()?;
+    socket.set_nodelay(true).to_con()?;
+
+    Ok(socket)
+}
+
+pub fn connect_to_client(
+    timeout: Duration,
+    client_ips: &[IpAddr],
+    port: u16,
+    send_buffer_bytes: SocketBufferSize,
+    recv_buffer_bytes: SocketBufferSize,
+) -> ConResult<TcpStream> {
+    let split_timeout = timeout / client_ips.len() as u32;
+
+    let mut res = alvr_common::try_again();
+    for ip in client_ips {
+        res = TcpStream::connect_timeout(&SocketAddr::new(*ip, port), split_timeout)
+            .handle_try_again();
+
+        if res.is_ok() {
+            break;
+        }
+    }
+    let socket = res?.into();
+
+    crate::set_socket_buffers(&socket, send_buffer_bytes, recv_buffer_bytes).ok();
+    socket.set_read_timeout(Some(timeout)).to_con()?;
+
+    let socket = TcpStream::from(socket);
+
+    socket.set_nodelay(true).to_con()?;
+
+    Ok(socket)
+}
+
+pub struct MultiplexedTcpWriter {
+    inner: TcpStream,
+}
+
+impl MultiplexedSocketWriter for MultiplexedTcpWriter {
+    fn payload_offset(&self) -> usize {
+        PACKET_PREFIX_SIZE
+    }
+
+    // `buffer` contains the payload offset by `payload_offset()`
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()> {
+        let payload_size = buffer.len() - PACKET_PREFIX_SIZE;
+
+        buffer[0..2].copy_from_slice(&stream_id.to_le_bytes());
+        buffer[2..6].copy_from_slice(&packet_index.to_le_bytes());
+        buffer[6..10].copy_from_slice(&(payload_size as u32).to_le_bytes());
+
+        self.inner.write_all(buffer)?;
+
+        Ok(())
+    }
+}
+
+struct InProgressPacket {
+    stream_id: u16,
+    packet_index: u32,
+    buffer: Vec<u8>,
+    buffer_size: usize,
+    cursor: usize,
+}
+
+pub struct MultiplexedTcpReader {
+    inner: Socket,
+    in_progress_packet: Option<InProgressPacket>,
+    used_buffers_poll_timeout: Duration,
+}
+
+impl MultiplexedSocketReader for MultiplexedTcpReader {
+    fn payload_offset(&self) -> usize {
+        PACKET_PREFIX_SIZE
+    }
+
+    fn recv(&mut self, stream_queues: &HashMap<u16, StreamRecvQueues>) -> ConResult {
+        let in_progress_packet = if let Some(packet) = &mut self.in_progress_packet {
+            packet
+        } else {
+            let mut prefix_bytes = [0u8; PACKET_PREFIX_SIZE];
+
+            loop {
+                let count = self
+                    .inner
+                    .peek(unsafe {
+                        &mut *(&mut prefix_bytes as *mut [u8] as *mut [MaybeUninit<u8>])
+                    })
+                    .handle_try_again()?;
+                if count == PACKET_PREFIX_SIZE {
+                    break;
+                }
+            }
```

Comment:
Honestly seems like a better decision to just return

Reply:
What do you mean?